src/test/java/spoon/test/filters/FilterTest.java

```diff
@@ -1124,4 +1125,17 @@ public void accept(CtType<?> ctType) {
 		// only one subtype remains unvisited
 		assertEquals(1, c2.counter);
 	}
+
+	@Test
+	public void testNameFilterWithGenericType() {
+		// contract: NameFilter of T should only return T elements
+
+		Launcher spoon = new Launcher();
+		spoon.addInputResource("./src/test/java/spoon/test/imports/testclasses/internal4/Constants.java");
+		spoon.buildModel();
+
+		CtType type = spoon.getFactory().Type().get(Constants.class);
+		List<CtMethod> ctMethods = type.getElements(new NameFilter<CtMethod>("CONSTANT"));
```

Comment:
keep only this test, add explanation, and ask @pvojtechovsky because the bug is somewhere in the query engine